File: 2021/day_2.rs
use std::fs;

fn load_data(filepath:&str) -> String {
  return fs::read_to_string(filepath).expect("Error!");
}

struct Position {
  h:i32,
  d:i32
}

struct ComplexPosition {
  h:i32,
  d:i32,
  a:i32,
}

struct Command {
  plane: String,
  distance: i32
}

fn first_star(data:&String) -> i32 {
  let mut pos = Position{h:0, d:0};

  for line in data.split("\n") {
    let cmd: Vec<&str> = line.split(" ").collect();
    let cmd = Command{plane:String::from(cmd[0]),
                      distance: cmd[1].parse().unwrap()};
    if cmd.plane == "forward" {
      pos.h += cmd.distance 
    } else if cmd.plane == "up" {
      pos.d -= cmd.distance
    } else {
      pos.d += cmd.distance
    }
  }

  return pos.h * pos.d;
}

fn second_star(data:&String) -> i32 {
  let mut pos = ComplexPosition{h:0, d:0, a:0};

  for line in data.split("\n") {
    let cmd: Vec<&str> = line.split(" ").collect();
    let cmd = Command{plane:String::from(cmd[0]),
                      distance: cmd[1].parse().unwrap()};
    if cmd.plane == "forward" {
      pos.h += cmd.distance;
      pos.d += pos.a * cmd.distance;
    } else if cmd.plane == "up" {
      pos.a -= cmd.distance;
    } else {
      pos.a += cmd.distance;
    }
  }

  return pos.h * pos.d;
}

fn main() {
  let data = load_data("input.txt");
  println!("First star: {}", first_star(&data));
  println!("Second star: {}", second_star(&data));
}

File: 2022/day_2.py
def load_input(filename):
  with open(filename) as f_obj:
    return [x.strip().split() for x in f_obj.readlines()]

def first_star(data):
    score = 0
    p = {'A': 1, 'B': 2, 'C': 3}
    m = { 'X' : 'A', 'Y': 'B', 'Z': 'C'}
    data = [(x[0], m[x[1]]) for x in data]
    for g in data:
        score += p[g[1]]
        if g[0] == g[1]:
            score += 3
        elif g in [('A', 'B'), ('B', 'C'), ('C', 'A')]:
            score += 6
        else:
            score += 0
    return score

def second_star(data):
    score = 0
    p = {'A': 1, 'B': 2, 'C': 3}
    for g in data:
        if g[1] == 'Y':
            score += 3
            score += p[g[0]]
        elif g[1] == 'Z':
            score += 6
            score += p[g[0]] + 1 if g[0] != 'C' else 1
        else:
            score += 0
            score += p[g[0]] - 1 if g[0] != 'A' else 3

    return score

if __name__ == "__main__":
  data = load_input("input.txt")
  print("First star:", first_star(data))
  print("Second star:", second_star(data))
